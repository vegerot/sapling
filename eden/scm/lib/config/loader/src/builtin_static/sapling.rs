/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use staticconfig::static_config;
use staticconfig::StaticConfig;

/// Config loaded only for the Sapling identity.
/// This config contains behavior changes when running "sl" vs "hg", so normally this is
/// _not_ where you want to add default config values.
pub static CONFIG: StaticConfig = static_config!("builtin:sapling" => r###"
[remotefilelog]
# Internally this will be overriden by dynamic config to be ~/.hgcache.
cachepath=~/.sl_cache

[committemplate]
changeset={if(desc, desc, emptymsg)}\n
 SL: Enter commit message.  Lines beginning with 'SL:' are removed.
 SL: {extramsg}
 SL: --
 SL: user: {author}\n{ifgt(parents|count, 1,
 "SL: merging:\n{parents % 'SL:   {node|short}: {desc|firstline}\n'}")
 }{if(currentbookmark,
 "SL: bookmark '{currentbookmark}'\n")}{
 filechanges}{
 if(advice, advice, defaultadvice)}

defaultadvice = SL: --
    SL: Consider onboarding Jellyfish in this repo to speed up your workflow.
    SL: Learn how at https://fburl.com/jf-onboard\n

filechangesplain={
 file_adds % "SL: added {file}\n"}{
 file_mods % "SL: changed {file}\n"}{
 file_dels % "SL: removed {file}\n"}{
 if(files, "", "SL: no files changed\n")}

[commands]
naked-default.in-repo=sl
"###);
