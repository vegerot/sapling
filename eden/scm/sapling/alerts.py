# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

import re
from typing import List, NamedTuple, Optional

from . import cmdutil, templater
from .i18n import _


class Alert(NamedTuple):
    severity: str
    title: str
    url: str
    description: str
    key: str
    show_in_isl: bool
    show_after_crashes_regex: Optional[re.Pattern]


def parse_alert(ui, key: str, raw_alert: dict) -> Optional[Alert]:
    # make sure the basic fields are present
    severity = raw_alert["severity"].strip()
    title = raw_alert["title"].strip()
    description = raw_alert["description"].strip()
    url = raw_alert["url"].strip()
    if not severity or not title or not description or not url:
        return None

    show_in_isl = raw_alert.get("show-in-isl", True)
    reg = raw_alert.get("show-after-crashes-regex")

    show_after_crashes_regex = None
    if reg:
        try:
            show_after_crashes_regex = re.compile(reg)
        except Exception:
            pass
    return Alert(
        key=key,
        severity=severity,
        title=title,
        url=url,
        description=description,
        show_in_isl=show_in_isl,
        show_after_crashes_regex=show_after_crashes_regex,
    )


def get_alerts(ui) -> List[Alert]:
    raw_alerts = {}
    for name, value in ui.configitems("alerts"):
        try:
            key, field = name.rsplit(".", 1)
        except Exception:
            continue
        if not raw_alerts.get(key):
            raw_alerts[key] = {}
        raw_alerts[key][field] = value

    alerts = []
    for key, raw_alert in raw_alerts.items():
        alert = parse_alert(ui, key, raw_alert)
        if alert:
            alerts.append(alert)

    return alerts


def severity_to_color(severity: str) -> str:
    if severity == "SEV 0" or severity == "UBN":
        return "alerts.critical"
    elif severity == "SEV 1":
        return "alerts.high"
    elif severity == "SEV 2":
        return "alerts.medium"
    elif severity == "SEV 3":
        return "alerts.low"
    elif severity == "SEV 4":
        return "alerts.advice"
    return "bold"


def print_alert(ui, alert: Alert):
    tmpl_str = ui.config("templatealias", "alerts")
    if tmpl_str is None:
        return
    template = templater.unquotestring(tmpl_str)
    ui.write_err(
        cmdutil.rendertemplate(
            ui,
            template,
            {
                "severity": alert.severity,
                "severity_color": severity_to_color(alert.severity),
                "url": alert.url,
                "title": alert.title,
                "description": alert.description,
            },
        )
    )


def print_active_alerts(ui):
    alerts = get_alerts(ui)

    for alert in alerts:
        print_alert(ui, alert)


def print_matching_alerts_for_exception(ui, crash: str):
    if ui.plain():
        return

    related_alerts = []

    alerts = get_alerts(ui)
    for alert in alerts:
        if alert.show_after_crashes_regex and alert.show_after_crashes_regex.search(
            crash
        ):
            related_alerts.append(alert)

    if not related_alerts:
        return

    ui.write_err(_("This crash may be related to an ongoing issue:\n"))
    for alert in related_alerts:
        print_alert(ui, alert)
