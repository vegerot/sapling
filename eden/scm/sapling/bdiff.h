/*
 * Portions Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

/*
 * Copyright Olivia Mackall <olivia@selenic.com> and others
 *
 * This software may be used and distributed according to the terms of
 * the GNU General Public License, incorporated herein by reference.
 */

#ifndef _HG_BDIFF_H_
#define _HG_BDIFF_H_

#include "eden/scm/sapling/compat.h"

struct bdiff_line {
  int hash, n, e;
  ssize_t len;
  const char* l;
};

struct bdiff_hunk;
struct bdiff_hunk {
  int a1, a2, b1, b2;
  struct bdiff_hunk* next;
};

int bdiff_splitlines(const char* a, ssize_t len, struct bdiff_line** lr);
int bdiff_diff(
    struct bdiff_line* a,
    int an,
    struct bdiff_line* b,
    int bn,
    struct bdiff_hunk* base);
void bdiff_freehunks(struct bdiff_hunk* l);

#endif
