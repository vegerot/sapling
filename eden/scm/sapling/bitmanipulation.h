/*
 * Portions Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

/*
 * Copyright Olivia Mackall <olivia@selenic.com> and others
 *
 * This software may be used and distributed according to the terms of
 * the GNU General Public License, incorporated herein by reference.
 */

#ifndef _HG_BITMANIPULATION_H_
#define _HG_BITMANIPULATION_H_

#include <string.h>

#include "eden/scm/sapling/compat.h"

#if defined(_MSC_VER)
/* Windows only supports little-endian platforms */
static inline uint64_t hg_be_u64(uint64_t x) {
  return _byteswap_uint64(x);
}
static inline uint32_t hg_be_u32(uint32_t x) {
  return _byteswap_ulong(x);
}
static inline uint16_t hg_be_u16(uint16_t x) {
  return _byteswap_ushort(x);
}
#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
static inline uint64_t hg_be_u64(uint64_t x) {
  return __builtin_bswap64(x);
}
static inline uint32_t hg_be_u32(uint32_t x) {
  return __builtin_bswap32(x);
}
static inline uint16_t hg_be_u16(uint16_t x) {
  return __builtin_bswap16(x);
}
#else
/* For completeness... */
static inline uint64_t hg_be_u64(uint64_t x) {
  return x;
}
static inline uint32_t hg_be_u32(uint32_t x) {
  return x;
}
static inline uint16_t hg_be_u16(uint16_t x) {
  return x;
}
#endif

static inline uint32_t getbe32(const char* c) {
  uint32_t value;
  memcpy(&value, c, sizeof(value));
  return hg_be_u32(value);
}

static inline uint16_t getbeuint16(const char* c) {
  uint16_t value;
  memcpy(&value, c, sizeof(value));
  return hg_be_u16(value);
}

static inline int16_t getbeint16(const char* c) {
  /*
   * Note: this code technically has undefined behavior for negative
   * values, although it's written in a way that the compiler and UBSAN
   * hopefully shouldn't complain about it.
   *
   * This relies on the platform using 2s-compliment representations for
   * signed integers.  This isn't guaranteed by the C standard, but is
   * true in practice for all modern platforms.
   */
  union {
    uint16_t unsignedvalue;
    int16_t signedvalue;
  } u;
  u.unsignedvalue = getbeuint16(c);
  return u.signedvalue;
}

static inline void putbe32(uint32_t x, char* c) {
  uint32_t v = hg_be_u32(x);
  memcpy(c, &v, sizeof(v));
}

static inline double getbefloat64(const char* c) {
  uint64_t n;
  double d;
  memcpy(&n, c, sizeof(n));
  n = hg_be_u64(n);
  memcpy(&d, &n, sizeof(n));
  return d;
}

#endif
