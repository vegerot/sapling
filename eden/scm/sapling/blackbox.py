# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

from bindings import blackbox as _blackbox

from . import prefork

events = _blackbox.events


def init(*args, **opts):
    if prefork.prefork:
        return
    _blackbox.init(*args, **opts)


def log(*args, **opts):
    if prefork.prefork:
        return
    _blackbox.log(*args, **opts)


sessions = _blackbox.sessions
sync = _blackbox.sync


def shortlist(listlike, count=None, limit=4):
    """Return a value that can be converted to Rust blackbox::event::ShortList"""
    shortlist = []
    for name in listlike:
        shortlist.append(name)
        if len(shortlist) > limit:
            break
    if count is None:
        count = len(listlike)
    return {"short_list": shortlist, "len": count}


class logblocked:
    def __new__(cls, op, seconds=None, name=None, ignorefast=False):
        """Log a "Blocked" event.

        If seconds is None, then this should be used as a context manager,
        and seconds will be calculated automatically. Otherwise, this
        function will log a "Blocked" event immediately.

        If name is not None, then additional name will be logged. This is
        useful for things like hook name where "op" is "hook", "name" is
        the actual hook name.

        If ignorefast is True, then a fast operation will be ignored.
        """
        if seconds is not None:
            # Non-context manager version
            millis = int(seconds * 1000)
            if not ignorefast or millis >= 10:
                log({"blocked": {"op": op, "duration_ms": millis, "name": name}})
            return None
        else:
            self = super(logblocked, cls).__new__(cls)
            self.op = op
            self.name = name
            self.ignorefast = ignorefast
            return self

    def __enter__(self):
        self.starttime = _timer()

    def __exit__(self, exctype, excval, exctb):
        seconds = _timer() - self.starttime
        millis = int(seconds * 1000)
        if not self.ignorefast or millis >= 10:
            log({"blocked": {"op": self.op, "duration_ms": millis, "name": self.name}})


def _timer():
    """util.timer"""
    from . import util  # avoid cycles

    globals()["_timer"] = util.timer  # replace self
    return util.timer()
