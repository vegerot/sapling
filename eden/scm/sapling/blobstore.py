# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

# blobstore.py - local blob storage

from __future__ import absolute_import

import hashlib

from . import error, util
from .i18n import _


class localblobstore:
    """A local blobstore.

    This blobstore is used both as a cache and as a staging area for large blobs
    to be uploaded to the remote blobstore.
    """

    def __init__(self, vfs, cachevfs):
        self.vfs = vfs
        self.cachevfs = cachevfs

    def write(self, oid, data):
        """Write blob to local blobstore."""
        contentsha256 = hashlib.sha256(data).hexdigest()
        if contentsha256 != oid:
            raise error.Abort(
                _("blobstore: sha256 mismatch (oid: %s, content: %s)")
                % (oid, contentsha256)
            )
        with self.vfs(oid, "wb", atomictemp=True) as fp:
            fp.write(data)

        # XXX: should we verify the content of the cache, and hardlink back to
        # the local store on success, but truncate, write and link on failure?
        if self.cachevfs and not self.cachevfs.exists(oid):
            self.vfs.linktovfs(oid, self.cachevfs)

    def read(self, oid):
        """Read blob from local blobstore."""
        if self.cachevfs and not self.vfs.exists(oid):
            self.cachevfs.linktovfs(oid, self.vfs)
        return self.vfs.read(oid)

    def has(self, oid):
        """Returns True if the local blobstore contains the requested blob,
        False otherwise."""
        return (self.cachevfs and self.cachevfs.exists(oid)) or self.vfs.exists(oid)

    def remove(self, oid):
        self.vfs.tryunlink(oid)

    def list(self):
        """Return a list of oids stored in this blobstore"""
        oids = []
        for entry in self.vfs.walk():
            oids += entry[-1]
        return sorted(oids)


class memlocal:
    """In-memory local blobstore for ad-hoc uploading/downloading without
    writing to the filesystem.

    Used by LFS (debuglfssingleupload and debuglfssingledownload).
    """

    def __init__(self):
        self._files = {}

    def write(self, oid, data):
        self._files[oid] = data

    def read(self, oid):
        return self._files[oid]

    def has(self, oid):
        return oid in self._files

    def vfs(self, oid, mode="r"):
        """wrapper for a "streaming" way to access a file

        Used by _gitlfsremote._basictransfer.
        """
        assert mode == "r"
        return util.stringio(self.read(oid))

    def remove(self, oid):
        if oid in self._files:
            del self._files[oid]

    def list(self):
        """Return a list of oids stored in this blobstore"""
        return list(sorted(self._files.keys()))


class unionstore:
    """A store which offers uniform access to in-memory store and local on-disk store."""

    def __init__(self, diskstore, memstore):
        self.diskstore = diskstore
        self.memstore = memstore

    def write(self, oid, data):
        self.diskstore.write(oid, data)

    def read(self, oid):
        if self.memstore.has(oid):
            return self.memstore.read(oid)
        else:
            return self.diskstore.read(oid)

    def has(self, oid):
        return self.memstore.has(oid) or self.diskstore.has(oid)

    def remove(self, oid):
        self.diskstore.remove(oid)
        self.memstore.remove(oid)

    def list(self):
        return list(sorted(set(self.diskstore.list() + self.memstore.list())))
