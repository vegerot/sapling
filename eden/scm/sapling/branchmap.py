# Portions Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

# branchmap.py - logic to computes, maintain and stores branchmap for local repo
#
# Copyright 2005-2007 Olivia Mackall <olivia@selenic.com>
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

from __future__ import absolute_import

from . import pycompat
from .node import nullid, nullrev


def updatecache(repo) -> None:
    # Just use a shortcut path that construct the branchcache directly.
    partial = branchcache()
    partial.update(repo, None)
    repo._branchcaches[None] = partial


class branchcache(dict):
    """A dict like object that hold branches heads cache.

    This cache is used to avoid costly computations to determine all the
    branch heads of a repo.

    The cache is serialized on disk in the following format:

    <tip hex node> <tip rev number> [optional filtered repo hex hash]
    <branch head hex node> <open/closed state> <branch name>
    <branch head hex node> <open/closed state> <branch name>
    ...

    The first line is used to check if the cache is still valid. If the
    branch cache is for a filtered repo view, an optional third hash is
    included that hashes the hashes of all filtered revisions.

    The open/closed state is represented by a single letter 'o' or 'c'.
    This field can be used to avoid changelog reads when determining if a
    branch head closes a branch or not.
    """

    def __init__(self, entries=(), tipnode=nullid, tiprev=nullrev, closednodes=None):
        super(branchcache, self).__init__(entries)
        self.tipnode = tipnode
        self.tiprev = tiprev
        # closednodes is a set of nodes that close their branch. If the branch
        # cache has been updated, it may contain nodes that are no longer
        # heads.
        if closednodes is None:
            self._closednodes = set()
        else:
            self._closednodes = closednodes

    def validfor(self, repo):
        """Is the cache content valid regarding a repo

        - False when cached tipnode is unknown or if we detect a strip.
        - True when cache is up to date or a subset of current repo."""
        try:
            return self.tipnode == repo.changelog.node(self.tiprev)
        except IndexError:
            return False

    def _branchtip(self, heads):
        """Return tuple with last open head in heads and false,
        otherwise return last closed head and true."""
        tip = heads[-1]
        closed = True
        for h in reversed(heads):
            if h not in self._closednodes:
                tip = h
                closed = False
                break
        return tip, closed

    def branchtip(self, branch):
        """Return the tipmost open head on branch head, otherwise return the
        tipmost closed head on branch.
        Raise KeyError for unknown branch."""
        return self._branchtip(self[branch])[0]

    def iteropen(self, nodes):
        return (n for n in nodes if n not in self._closednodes)

    def branchheads(self, branch, closed=False):
        heads = self[branch]
        if not closed:
            heads = list(self.iteropen(heads))
        return heads

    def iterbranches(self):
        for bn, heads in pycompat.iteritems(self):
            yield (bn, heads) + self._branchtip(heads)

    def copy(self) -> "branchcache":
        """return an deep copy of the branchcache object"""
        return branchcache(self, self.tipnode, self.tiprev, self._closednodes)

    def update(self, repo, revgen):
        """Given a branchhead cache, self, that may have extra nodes or be
        missing heads, and a generator of nodes that are strictly a superset of
        heads missing, this function updates self to be correct.
        """
        # Behave differently if the cache is disabled.
        cl = repo.changelog
        tonode = cl.node

        if self.tiprev == len(cl) - 1 and self.validfor(repo):
            return

        # Since we have no branches, the default branch heads are equal to
        # repo.headrevs(). Note: repo.headrevs() is already sorted and it may
        # return -1.
        branchheads = [i for i in repo.headrevs(reverse=False) if i >= 0]

        if not branchheads:
            if "default" in self:
                del self["default"]
            tiprev = -1
        else:
            self["default"] = [tonode(rev) for rev in branchheads]
            tiprev = branchheads[-1]
        self.tipnode = cl.node(tiprev)
        self.tiprev = tiprev
        repo.ui.log("branchcache", "perftweaks updated branch cache\n")
