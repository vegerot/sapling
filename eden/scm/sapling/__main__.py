# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

from __future__ import absolute_import

import sapling


sapling.run()
