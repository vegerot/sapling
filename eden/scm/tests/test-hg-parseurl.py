from __future__ import absolute_import, print_function

from sapling import hg


def testparse(url):
    print("%s" % hg.parseurl(url))


testparse("http://example.com/no/anchor")
testparse("http://example.com/an/anchor#foo")
testparse("http://example.com/no/anchor/branches")
testparse("http://example.com/an/anchor/branches#bar")
testparse("http://example.com/an/anchor/branches-None#foo")
testparse("http://example.com/")
testparse("http://example.com")
testparse("http://example.com#foo")
