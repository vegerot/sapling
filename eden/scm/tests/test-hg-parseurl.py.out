http://example.com/no/anchor
http://example.com/an/anchor
http://example.com/no/anchor/branches
http://example.com/an/anchor/branches
http://example.com/an/anchor/branches-None
http://example.com/
http://example.com/
http://example.com/
